### PR TITLE
all: implement minimal repo listing

### PR DESCRIPTION
**Pull request by tsenart**
This change set implements minimal repo listing across our Searcher implementations.

This is meant to reduce memory utilization in Sourcegraph's frontend, which holds on to the response of the List method for several use cases.

Some tests are failing with an unexpected negative document count that I think might point to a bug in stats calculation here: https://github.com/sourcegraph/zoekt/blob/099907aa573abce45ce534364586ac25a6820a07/indexdata.go#L178

Or am I breaking an invariant that an index always contains at least one document?

### CODE REVIEW

**Comment by keegancsmith** [comment]
Will review today. FYI the stats where wrong, we were following up with a fix. The fix is here https://github.com/sourcegraph/zoekt/pull/98

File: api.go [hunk 1]
```diff
@@ -321,24 +328,43 @@ type RepoListEntry struct {
...
+
 // RepoList holds a set of Repository metadata.
 type RepoList struct {
+	// Full response to a List request. Returned when ListOptions.Minimal is false.
 	Repos   []*RepoListEntry
+
 	Crashes int
+
+	// Minimal response to a List request. Returned when ListOptions.Minimal is true.
+	Minimal map[uint32]*MinimalRepoListEntry
```
Comment (keegancsmith) [inline api.go:359]: why a map? Can't we just use a list? list uses less mem + more efficient to marshal/unmarshal. Also get the benefit of simpler aggregation code, you can just use append.

File: eval.go [hunk 2]
```diff
@@ -450,34 +452,65 @@ func (d *indexData) List(ctx context.Context, q query.Q) (rl *RepoList, err erro
...
-	l := &RepoList{}
-	if include {
-		l.Repos = make([]*RepoListEntry, 0, len(d.repoListEntry))
-		for _, rle := range d.repoListEntry {
-			l.Repos = append(l.Repos, &rle)
+		if !ok {
+			continue
+		}
+
+		if id := rle.Repository.ID(); id != 0 && minimal {
```
Comment (keegancsmith) [inline eval.go]: it feels weird we need to parse the ID each time. I'm thinking we add the ID field and on unmarshalling repository from the shard we set the ID? Additionally we just start to set ID on Repository?

File: shards/eval.go [hunk 3]
```diff
@@ -95,7 +97,7 @@ func (s *typeRepoSearcher) eval(ctx context.Context, q query.Q) (query.Q, error)
 		}
 
 		var rl *zoekt.RepoList
-		rl, err = s.Streamer.List(ctx, rq.Child)
```
Comment (keegancsmith) [inline shards/eval.go:98]: for future, but minimal would be useful here since we are calling list here to find which repos to search.

**Review by keegancsmith — APPROVED** [review]
LGTM. The only thing I don't really like is the use of a map. You'll notice a lack of maps in the API. I'm guessing you want a map to simplify work on Sourcegraph's side? I'd still rather stick to slices TBH.

File: api.go (hunk 1 above)
Comment (tsenart) [reply api.go:359]: Both in Sourcergaph's side and in shardedSearcher we need to use maps for deduplication and fast lookups. Even for the existing normal `RepoListEntry` we already create an intermediate `map[string]*RepoListEntry` for deduping. On Sourceraph's side, we convert all of that to map again, so any gains in marshalling performance are offset by that.

How much more memory does a map use in Gob encoding?

File: eval.go (hunk 2 above)
Comment (tsenart) [reply eval.go]: Sounds good to me. Wasn't confident we wanted to do that, thanks for confirming!

File: shards/eval.go (hunk 3 above)
Comment (tsenart) [reply shards/eval.go:98]: We'd have to make RepoSet work with IDs at that point. Or introduce a RepoIDSet. What's the approach you have in mind?

File: eval.go (hunk 2 above)
Comment (tsenart) [reply eval.go]: I'll do a follow up PR that sets the proper ID field in the repo metadata to limit the scope of this PR.

Comment (tsenart) [reply eval.go:503]: I'll do a follow up PR that sets the proper ID field in the repo metadata to limit the scope of this PR.

Comment (tsenart) [reply eval.go]: https://github.com/sourcegraph/zoekt/pull/99/commits/8e34ee2e879933487ad02d0a2ef03239f565ed3d

Comment (tsenart) [reply eval.go]: https://github.com/sourcegraph/zoekt/pull/99/commits/8e34ee2e879933487ad02d0a2ef03239f565ed3d

File: api.go (hunk 1 above)
Comment (keegancsmith) [reply api.go:359]: Cool, after more chatting we are gonna forward with this. It shouldn't make that much of a difference, so it's more aesthetic. Additionally the benefit of fast lookups is great on the client side.

File: eval.go (hunk 2 above)
Comment (keegancsmith) [reply eval.go]: nice!

File: shards/eval.go (hunk 3 above)
Comment (keegancsmith) [reply shards/eval.go:98]: yup. We mainly use RepoBranches so that would also need be needed.

File: shards/shards_test.go [hunk 4]
```diff
@@ -284,6 +287,111 @@ func TestUnloadIndex(t *testing.T) {
...
+			}
+
+			sort.Slice(res.Repos, func(i, j int) bool {
+				return res.Repos[i].Repository.Name < res.Repos[j].Repository.Name
+			})
+
+			ignored := []cmp.Option{
+				cmpopts.EquateEmpty(),
+				cmpopts.IgnoreFields(zoekt.RepoListEntry{}, "IndexMetadata"),
+				cmpopts.IgnoreFields(zoekt.RepoStats{}, "IndexBytes", "Documents"),
```
Comment (keegancsmith) [inline shards/shards_test.go]: @tsenart I merged in https://github.com/sourcegraph/zoekt/pull/98 so you can remove this ignore after rebasing.

File: api.go (hunk 1 above)
Comment (tsenart) [reply api.go:359]: FYI @keegancsmith, map is slightly smaller in serialized size but a little slower and does allocs. Could likely use a custom Gob encoding function: https://github.com/sourcegraph/zoekt/pull/99/commits/72bcfba482eec95719af3a13e5c1e40628fde080

**Pull request merged**